.clang-format

```diff
@@ -0,0 +1,22 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+---
+BasedOnStyle: Google
+ColumnLimit: 90
+DerivePointerAlignment: false
+IncludeBlocks: Preserve
```

Comment:
I personally find the output of 'Regroup' more readable

Reply:
Hi, @wgtmac @pitrou @raulcd, what's your opinions on this one. I copied this from arrow project, and I see ORC also has the same setting, so I myself prefer this as is.

Reply:
Other than another project uses this, is there any reason you prefer this setting? With 'Regroup' the includes are more structured and organised. What are the pros for 'Preserve'?

Reply:
> Other than another project uses this, is there any reason you prefer this setting? With 'Regroup' the includes are more structured and organised. What are the pros for 'Preserve'?

We use Preserve in our own projects, developers who add the header files should give them a reasonable position.

Reply:
Using Preserve allows us to have separate include blocks for 1) stdlib includes 2) third-party includes 3) own project includes.

However, it seems the same effect could be achieved with Regroup and a carefully-defined IncludeCategories?

Reply:
even though I don't have much experience with it, `Regroup` seems like a good approach to me

Reply:
The doc says `IncludeCategories` can be used with any `IncludeBlocks` strategy, there is a additional field `SortPriority` which can be used with `Regroup`[0].  Here is the setting from Impala, do you think this works for us?

```
IncludeCategories:
- Priority: 3
  Regex: '^<.*/' # Like <boost/lexical_cast.hpp>
- Priority: 1
  Regex: '^<.*\.' # Like <fcntl.h>
- Priority: 2
  Regex: '^<' # Like <vector>
- Priority: 4
  Regex: '^"' # Like "util/auth-util.h"
```

[0] https://rocm.docs.amd.com/projects/llvm-project/en/latest/LLVM/clang/html/ClangFormatStyleOptions.html#includecategories

Reply:
As long as it allows dissociating stdlib, third-party, and own project imports, then I'm fine with it.

Reply:
I plan to make the following change about the `IncludeBlocks`

```
-IncludeBlocks: Preserve
+IncludeBlocks: Regroup
+IncludeCategories:
+- Regex: '^<.*/' # Like <boost/lexical_cast.hpp>
+  Priority: 3
+- Regex: '^<.*\.' # Like <fcntl.h>
+  Priority: 1
+- Regex: '^<' # Like <vector>
+  Priority: 2
+- Regex: '^"' # Like "util/auth-util.h"
+  Priority: 4
```

Before I push a new commit, I'd like to explain why I think this makes sense.

I did a experiment, `Regroup` seems to be the default setting if we don't specify a `IncludeBlocks` setting, I put it here explicitly because I think it's more clear for the readers. With the `IncludeCategories` settings, the following:

```
#include <iostream>
#include <string>

#include <boost/lexical_cast.hpp>

#include "util/auth-util.h"

#include <vector>

#include <fcntl.h>

#include "demo_table.h"
```

goes to:

```
#include "demo_table.h"

#include <fcntl.h>

#include <iostream>
#include <string>
#include <vector>

#include <boost/lexical_cast.hpp>

#include "util/auth-util.h"
```

It's slightly different from Impala's setting because I think putting the `Regex` as the first line of each rule is more readable, "If match this regex, then goes priority xxx".

What do you about this proposal? @gaborkaszab @pitrou @Fokko @wgtmac @raulcd

Reply:
Seems reasonable for me. Thanks!

Reply:
Pushed the change, thanks for your reaction.